### PR TITLE
Convert libcudf hashing benchmarks to nvbench 

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Converts gbench hashing benchmark to use nvbench instead. This makes it easier to maintain and to customize benchmark runs.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/hashing/hash.cpp [hunk 1]
```diff
@@ -15,47 +15,47 @@
...
+  if (hash_name == "murmur32") {
+    state.exec(nvbench::exec_tag::sync, [&](nvbench::launch& launch) {
+      auto result = cudf::hashing::murmur_hash3_32(data->view());
+    });
+  } else if (hash_name == "md5") {
+    state.exec(nvbench::exec_tag::sync,
+               [&](nvbench::launch& launch) { auto result = cudf::hashing::md5(data->view()); });
+  } else if (hash_name == "spark_murmur32") {
+    state.exec(nvbench::exec_tag::sync, [&](nvbench::launch& launch) {
+      auto result = cudf::hashing::murmur_hash3_32(data->view());
```
Comment (bdice) [inline cpp/benchmarks/hashing/hash.cpp:50]:
```suggestion
      auto result = cudf::hashing::spark_murmur_hash3_32(data->view());
```

File: cpp/benchmarks/hashing/hash.cpp [hunk 2]
```diff
@@ -15,47 +15,47 @@
...
-HASH_BENCHMARK_DEFINE(HASH_MURMUR3, nulls)
-HASH_BENCHMARK_DEFINE(HASH_SPARK_MURMUR3, nulls)
-HASH_BENCHMARK_DEFINE(HASH_MD5, nulls)
-
-HASH_BENCHMARK_DEFINE(HASH_MURMUR3, no_nulls)
-HASH_BENCHMARK_DEFINE(HASH_SPARK_MURMUR3, no_nulls)
-HASH_BENCHMARK_DEFINE(HASH_MD5, no_nulls)
+NVBENCH_BENCH(bench_hash)
+  .set_name("hashing")
+  .add_int64_axis("num_rows", {16384, 65536, 262144, 1048576, 4194304, 16777216})
```
Comment (bdice) [inline cpp/benchmarks/hashing/hash.cpp:59]: Do we need this many different values? One small and one large might suffice.

File: cpp/benchmarks/hashing/hash.cpp [hunk 3]
```diff
@@ -15,47 +15,47 @@
...
+  .add_int64_axis("nulls", {0, 1})
+  .add_string_axis("hash_name", {"murmur32", "md5", "spark_murmur32"});
```
Comment (bdice) [inline cpp/benchmarks/hashing/hash.cpp:61]: Match these names to the hashing API function name. We’ve been very loose with the spelling of MurmurHash3 and I’d like to standardize that through the codebase.

Comment (davidwendt) [reply cpp/benchmarks/hashing/hash.cpp:61]: I was just trying to keep the column name's smaller.
```
# Benchmark Results

## hashing

### [0] NVIDIA RTX A6000

| num_rows | nulls |       hash_name       | Samples |  CPU Time  | Noise  |  GPU Time  | Noise |
|----------|-------|-----------------------|---------|------------|--------|------------|-------|
|    65536 |     0 |       murmur_hash3_32 |  17472x |  33.970 us | 19.20% |  28.619 us | 2.50% |
| 16777216 |     0 |       murmur_hash3_32 |    528x | 953.777 us |  0.60% | 948.329 us | 0.16% |
|    65536 |     1 |       murmur_hash3_32 |   8400x |  64.964 us |  9.11% |  59.593 us | 1.48% |
| 16777216 |     1 |       murmur_hash3_32 |    495x |   1.016 ms |  0.56% |   1.011 ms | 0.17% |
|    65536 |     0 |                   md5 |   4624x | 113.625 us |  5.44% | 108.302 us | 2.34% |
| 16777216 |     0 |                   md5 |     46x |  10.909 ms |  0.48% |  10.904 ms | 0.48% |
|    65536 |     1 |                   md5 |   4528x | 115.895 us |  5.17% | 110.544 us | 1.83% |
| 16777216 |     1 |                   md5 |    720x |  11.172 ms |  0.63% |  11.166 ms | 0.63% |
|    65536 |     0 | spark_murmur_hash3_32 |  17712x |  33.576 us | 19.31% |  28.231 us | 4.07% |
| 16777216 |     0 | spark_murmur_hash3_32 |    633x | 795.458 us |  0.70% | 789.988 us | 0.09% |
|    65536 |     1 | spark_murmur_hash3_32 |   8496x |  64.296 us |  9.21% |  58.938 us | 1.61% |
| 16777216 |     1 | spark_murmur_hash3_32 |   1751x | 863.646 us |  0.81% | 858.156 us | 0.50% |

```

File: cpp/benchmarks/hashing/hash.cpp [hunk 4]
```diff
@@ -15,47 +15,47 @@
...
+  auto const num_rows  = static_cast<cudf::size_type>(state.get_int64("num_rows"));
+  auto const has_nulls = static_cast<bool>(state.get_int64("nulls"));
+  auto const hash_name = state.get_string("hash_name");
+
+  auto const data =
+    create_random_table({cudf::type_id::INT64, cudf::type_id::STRING}, row_count{num_rows});
+  if (!has_nulls) {
     data->get_column(0).set_null_mask(rmm::device_buffer{}, 0);
+    data->get_column(1).set_null_mask(rmm::device_buffer{}, 0);
+  }
```
Comment (karthikeyann) [inline cpp/benchmarks/hashing/hash.cpp:36]:
```suggestion
  data_profile profile;
  if (!has_nulls) {
    profile.set_null_probability(0.0);
  }
  auto const data =
    create_random_table({cudf::type_id::INT64, cudf::type_id::STRING}, row_count{num_rows});
```
or consider using `data_profile_builder`

File: cpp/benchmarks/hashing/hash.cpp [hunk 5]
```diff
@@ -15,47 +15,44 @@
...
-HASH_BENCHMARK_DEFINE(HASH_MD5, nulls)
-
-HASH_BENCHMARK_DEFINE(HASH_MURMUR3, no_nulls)
-HASH_BENCHMARK_DEFINE(HASH_SPARK_MURMUR3, no_nulls)
-HASH_BENCHMARK_DEFINE(HASH_MD5, no_nulls)
+NVBENCH_BENCH(bench_hash)
+  .set_name("hashing")
+  .add_int64_axis("num_rows", {65536, 16777216})
+  .add_float64_axis("nulls", {0.0, 0.1})
+  .add_string_axis("hash_name", {"murmur_hash3_32", "md5", "spark_murmur_hash3_32"});
```
Comment (bdice) [inline cpp/benchmarks/hashing/hash.cpp:58]: Maybe we can align the name with the discussions in #13681?
```suggestion
  .add_string_axis("hash_name", {"murmurhash3_x86_32", "md5", "spark_murmur_hash3_32"});
```

**Review by bdice — APPROVED** [review]

File: cpp/benchmarks/hashing/hash.cpp (hunk 5 above)
Comment (davidwendt) [reply cpp/benchmarks/hashing/hash.cpp:58]: Happy to do the once #13681 is merged.

**Review by karthikeyann — APPROVED** [review]
LGTM 👍

**Pull request merged**